File: Bukkit/src/main/java/me/mrkirby153/kcutils/ItemFactory.java
package me.mrkirby153.kcutils;

import org.bukkit.ChatColor;
import org.bukkit.Material;
import org.bukkit.enchantments.Enchantment;
import org.bukkit.enchantments.EnchantmentTarget;
import org.bukkit.enchantments.EnchantmentWrapper;
import org.bukkit.inventory.ItemFlag;
import org.bukkit.inventory.ItemStack;
import org.bukkit.inventory.meta.ItemMeta;

import java.lang.reflect.Field;
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;

/**
 * Utility class for creating items
 */
public class ItemFactory {

    public static final Enchantment ENCHANTMENT_GLOWING = new GlowEnchantment();

    private Material material;
    private short data;
    private String name;
    private int amount;
    private boolean unbreakable;
    private List<String> lore;
    private List<ItemFlag> flags;
    private HashMap<Enchantment, Integer> enchantments;


    public ItemFactory(Material material) {
        this.material = material;
        this.data = 0;
        this.amount = 1;
        this.unbreakable = false;
        this.lore = new ArrayList<>();
        this.flags = new ArrayList<>();
        this.enchantments = new HashMap<>();
    }

    /**
     * Sets the item amount
     *
     * @param amount The item amount
     * @return The factory
     */
    public ItemFactory amount(int amount) {
        this.amount = amount;
        return this;
    }

    /**
     * Turns the item factory into an Item Stack
     *
     * @return An {@link ItemStack}
     */
    public ItemStack construct() {
        ItemStack stack = new ItemStack(material, amount, data);
        ItemMeta meta = stack.getItemMeta();
        if (name != null)
            meta.setDisplayName(ChatColor.RESET + name);
        meta.setLore(lore);
        meta.setUnbreakable(unbreakable);
        meta.addItemFlags(flags.toArray(new ItemFlag[0]));
        enchantments.forEach((e, l) -> meta.addEnchant(e, l, true));
        stack.setItemMeta(meta);

        return stack;
    }

    /**
     * Sets the item's data
     *
     * @param data The data
     * @return The factory
     */
    public ItemFactory data(int data) {
        this.data = (short) data;
        return this;
    }

    /**
     * Adds an item's enchantment
     *
     * @param enchantment The enchantment
     * @param level       The level
     * @return The factory
     */
    public ItemFactory enchantment(Enchantment enchantment, int level) {
        this.enchantments.put(enchantment, level);
        return this;
    }

    /**
     * Adds flags to an item
     *
     * @param flags The flags
     * @return The factory
     */
    public ItemFactory flags(ItemFlag... flags) {
        Collections.addAll(this.flags, flags);
        return this;
    }

    /**
     * Makes the item glow
     *
     * @return The factory
     */
    public ItemFactory glowing() {
        this.enchantment(ENCHANTMENT_GLOWING, 1);
        return this;
    }

    /**
     * Adds lore to an item
     *
     * @param lore The lore
     * @return The factory
     */
    public ItemFactory lore(String... lore) {
        for (String l : lore) {
            this.lore.add(ChatColor.RESET + l);
        }
        return this;
    }

    /**
     * Sets the item's name
     *
     * @param name The name
     * @return The factory
     */
    public ItemFactory name(String name) {
        this.name = name;
        return this;
    }

    /**
     * Makes the item unbreakable
     *
     * @return The factory
     */
    public ItemFactory unbreakable() {
        this.unbreakable = true;
        return this;
    }

    public static class GlowEnchantment extends EnchantmentWrapper {

        public GlowEnchantment() {
            super(120);
            register();
        }

        @Override
        public boolean canEnchantItem(ItemStack item) {
            return true;
        }

        @Override
        public boolean conflictsWith(Enchantment other) {
            return false;
        }

        @Override
        public EnchantmentTarget getItemTarget() {
            return null;
        }

        @Override
        public int getMaxLevel() {
            return 1;
        }

        @Override
        public String getName() {
            return "NullEnchantment";
        }

        @Override
        public int getStartLevel() {
            return 1;
        }

        private void register() {
            try {
                // Register enchantment
                Field f = Enchantment.class.getDeclaredField("acceptingNew");
                f.setAccessible(true);
                f.setAccessible(true);
                boolean previous = f.getBoolean(null);
                f.set(null, true);
                if (Enchantment.getByName("NullEnchantment") == null)
                    Enchantment.registerEnchantment(this);

                f.setBoolean(null, previous);
            } catch (IllegalAccessException | NoSuchFieldException e) {
                e.printStackTrace();
            }
        }
    }
}

File: Bukkit/src/main/kotlin/me/mrkirby153/kcutils/ItemFactory.kt
package me.mrkirby153.kcutils

import net.kyori.adventure.text.Component
import org.bukkit.Material
import org.bukkit.NamespacedKey
import org.bukkit.enchantments.Enchantment
import org.bukkit.enchantments.EnchantmentWrapper
import org.bukkit.inventory.ItemFlag
import org.bukkit.inventory.ItemStack
import org.bukkit.inventory.meta.Damageable

/**
 * Utility class for creating items
 */
class ItemFactory(private val material: Material) {
    private var damage: Int = 0
    private var name: Component? = null
    private var amount: Int = 1
    private var unbreakable: Boolean = false
    private val lore: MutableList<Component> = mutableListOf()
    private val flags: MutableList<ItemFlag> = mutableListOf()
    private val enchantments: MutableMap<Enchantment, Int> = mutableMapOf()

    /**
     * Sets the item amount
     *
     * @param amount The item amount
     *
     * @return The factory
     */
    fun amount(amount: Int): ItemFactory {
        this.amount = amount
        return this
    }

    /**
     * Turns the item factory into an Item Stack
     *
     * @return An [ItemStack]
     */
    fun construct(): ItemStack {
        val stack = ItemStack(material, amount)
        val meta = stack.itemMeta!!
        if (name != null) {
            meta.displayName(name)
        }
        if (meta is Damageable) {
            meta.damage = this.damage
        }
        meta.lore()
        meta.isUnbreakable = unbreakable
        meta.addItemFlags(*flags.toTypedArray())
        enchantments.forEach { (e, l) -> meta.addEnchant(e, l, true) }
        stack.itemMeta = meta

        return stack
    }

    /**
     * Sets the item's damage
     *
     * @param damage The damage
     *
     * @return The factory
     */
    fun damage(damage: Int): ItemFactory {
        this.damage = damage
        return this
    }

    /**
     * Adds an item's enchantment
     *
     * @param enchantment The enchantment
     * @param level       The level
     *
     * @return The factory
     */
    fun enchantment(enchantment: Enchantment, level: Int): ItemFactory {
        this.enchantments[enchantment] = level
        return this
    }

    /**
     * Adds flags to an item
     *
     * @param flags The flags
     *
     * @return The factory
     */
    fun flags(vararg flags: ItemFlag): ItemFactory {
        this.flags.addAll(flags)
        return this
    }

    /**
     * Makes the item glow
     *
     * @return The factory
     */
    fun glowing(): ItemFactory {
        this.enchantment(ENCHANTMENT_GLOWING, 1)
        return this
    }

    /**
     * Adds lore to an item
     *
     * @param lore The lore
     *
     * @return The factory
     */
    fun lore(vararg lore: String): ItemFactory {
        for (l in lore) {
            this.lore.add(Component.text(l))
        }
        return this
    }

    /**
     * Sets the item's name
     *
     * @param name The name
     *
     * @return The factory
     */
    fun name(name: String): ItemFactory {
        this.name = Component.text(name)
        return this
    }

    fun name(name: Component): ItemFactory {
        this.name = name
        return this
    }

    /**
     * Makes the item unbreakable
     *
     * @return The factory
     */
    fun unbreakable(): ItemFactory {
        this.unbreakable = true
        return this
    }

    /**
     * An enchantment for making an item glow
     */
    class GlowEnchantment : EnchantmentWrapper("glow") {
        init {
            register()
        }

        override fun canEnchantItem(item: ItemStack): Boolean {
            return true
        }

        override fun conflictsWith(other: Enchantment): Boolean {
            return false
        }

        override fun getMaxLevel(): Int {
            return 1
        }

        @Deprecated("Deprecated in Java")
        override fun getName(): String {
            return "NullEnchantment"
        }

        override fun getStartLevel(): Int {
            return 1
        }

        private fun register() {
            try {
                // Register enchantment
                val f = Enchantment::class.java.getDeclaredField("acceptingNew")
                f.isAccessible = true
                f.isAccessible = true
                val previous = f.getBoolean(null)
                f.set(null, true)
                if (Enchantment.getByKey(NamespacedKey.minecraft("glow")) == null)
                    Enchantment.registerEnchantment(this)

                f.setBoolean(null, previous)
            } catch (e: IllegalAccessException) {
                e.printStackTrace()
            } catch (e: NoSuchFieldException) {
                e.printStackTrace()
            }

        }
    }

    companion object {

        /**
         * The glowing enchantment -- An enchantment with no effects whose sole purpose is to make the item glow
         */
        val ENCHANTMENT_GLOWING: Enchantment = GlowEnchantment()
    }
}
